### PR TITLE
nim bindings

### PR DESCRIPTION
**Pull request by arnetheduck**

### CODE REVIEW

File: hashtree_abi.nim [hunk 1]
```diff
@@ -0,0 +1,29 @@
+# hashtree nim bindings
+# Copyright (c) 2024 Status Research & Development GmbH
+# Licensed and distributed under either of
+#   * MIT license (license terms in the root directory or at https://opensource.org/licenses/MIT).
+#   * Apache v2 license (license terms in the root directory or at https://www.apache.org/licenses/LICENSE-2.0).
+# at your option. This file may not be copied, modified, or distributed except according to those terms.
+
```
Comment (etan-status) [inline hashtree_abi.nim:7]: `{.push raises: [].}` here maybe a bit more consistent than adding it to each proc/func

File: hashtree_abi.nimble [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+packageName   = "hashtree_abi"
+version       = "0.1.0"
+author        = "Prysmatic labs, Status Research and Development GmbH"
+description   = "Low-level ABI package for hashtree, a SHA256 implementation specialized for merkle trees and 64-byte chunks"
+license       = "MIT"
+installDirs   = @["src"]
+installFiles = @["hashtree_abi.nim"]
+
+requires "nim >= 1.2.0"
```
Comment (etan-status) [inline hashtree_abi.nimble:11]:
```suggestion
requires "nim >= 1.6.0"
```
or, alternatively, add 1.2.0 tests to the CI as well.

File: tests/hashtree.nim [hunk 3]
```diff
@@ -0,0 +1,6 @@
+import ../hashtree_abi
```
Comment (etan-status) [inline tests/hashtree.nim:1]: copyright missing

File: hashtree_abi.nimble [hunk 4]
```diff
@@ -0,0 +1,11 @@
+mode = ScriptMode.Verbose
```
Comment (etan-status) [inline hashtree_abi.nimble:1]: copyright missing, they also have one in src/Makefile

File: hashtree_abi.nimble [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+
+packageName   = "hashtree_abi"
+version       = "0.1.0"
```
Comment (etan-status) [inline hashtree_abi.nimble:4]: Note to Prysmatic: Nimble packages are released by tagging the repo with `v0.1.0` in this case. That's what gets downloaded when `nimble install` is requested for `0.1.0`. Right now, there don't seem to be any releases yet, but there is a tiny maintenance burden of keeping the latest tag in sync with what's written here.

File: hashtree_abi.nim [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+  {.compile: srcDir & "sha256_armv8_neon_x1.S".}
+  {.compile: srcDir & "sha256_armv8_neon_x4.S".}
+
+elif defined(amd64):
+  {.compile: srcDir & "sha256_avx_x1.S".}
+  {.compile: srcDir & "sha256_avx_x4.S".}
+  {.compile: srcDir & "sha256_avx_x8.S".}
+  {.compile: srcDir & "sha256_avx_x16.S".}
+  {.compile: srcDir & "sha256_shani.S".}
+  {.compile: srcDir & "sha256_sse_x1.S".}
```
Comment (etan-status) [inline hashtree_abi.nim:24]: Fails for me on macOS Intel.

```sh
% sysctl hw.model
hw.model: MacBookPro16,1
```

```sh
% sw_vers
ProductName:		macOS
ProductVersion:		14.3.1
BuildVersion:		23D60
```

```zsh
% nim c -r tests/hashtree.nim
Hint: used config file '/usr/local/Cellar/nim/2.0.2/nim/config/nim.cfg' [Conf]
Hint: used config file '/usr/local/Cellar/nim/2.0.2/nim/config/config.nims' [Conf]
...................................................................................................................
CC: sha256_avx_x1
CC: sha256_avx_x4
CC: sha256_avx_x8
CC: sha256_avx_x16
CC: sha256_shani
CC: sha256_sse_x1
/Users/etan/Documents/Repos/hashtree/src/sha256_shani.S:27:17: error: unexpected token in '.section' directive
.section .rodata
                ^
/Users/etan/Documents/Repos/hashtree/src/sha256_shani.S:112:1: error: unknown directive
.type hashtree_sha256_shani_x2,%function
^
/Users/etan/Documents/Repos/hashtree/src/sha256_shani.S:130:2: error: invalid operand for instruction
 movdqa SHUF_MASK, [rip + .LPSHUFFLE_BYTE_FLIP_MASK]
 ^
/Users/etan/Documents/Repos/hashtree/src/sha256_shani.S:131:2: error: invalid operand for instruction
 lea SHA256CONSTANTS,[rip + .LK256]
 ^
/Users/etan/Documents/Repos/hashtree/src/sha256_shani.S:132:2: error: invalid operand for instruction
 lea SHA256PADDING,[rip + .LPADDING]
 ^
... (many more lines)
```

File: tests/hashtree.nim [hunk 7]
```diff
@@ -0,0 +1,6 @@
...
+
+var data: array[64, byte]
+hashtree_hash(addr data[0], addr data[0], 1)
+
+doAssert data[0] == 245
```
Comment (etan-status) [inline tests/hashtree.nim:6]: quite minimal, but tests the core funtionality of the binding. more exhaustive tests should probably not be nim specific, so is ok

File: hashtree_abi.nimble [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+author        = "Prysmatic labs, Status Research and Development GmbH"
+description   = "Low-level ABI package for hashtree, a SHA256 implementation specialized for merkle trees and 64-byte chunks"
+license       = "MIT"
+installDirs   = @["src"]
+installFiles = @["hashtree_abi.nim"]
```
Comment (etan-status) [inline hashtree_abi.nimble:9]:
```suggestion
installFiles  = @["hashtree_abi.nim"]
```

File: hashtree_abi.nim (hunk 6 above)
Comment (arnetheduck) [reply hashtree_abi.nim:24]: this is expected, the library does not yet support mac

File: hashtree_abi.nim [hunk 9]
```diff
@@ -0,0 +1,29 @@
+# hashtree nim bindings
+# Copyright (c) 2024 Status Research & Development GmbH
+# Licensed and distributed under either of
+#   * MIT license (license terms in the root directory or at https://opensource.org/licenses/MIT).
+#   * Apache v2 license (license terms in the root directory or at https://www.apache.org/licenses/LICENSE-2.0).
```
[the remark this answers is not in the document]
Comment (arnetheduck) [reply hashtree_abi.nim]: a300d24da3482bc1a548467654b6bc1b98966fbc

File: hashtree_abi.nim (hunk 1 above)
Comment (arnetheduck) [reply hashtree_abi.nim:8]: a300d24da3482bc1a548467654b6bc1b98966fbc - on imported functions, push doesn't work too well, but we can avoid some copypaste at least

File: tests/hashtree.nim (hunk 3 above)
Comment (arnetheduck) [reply tests/hashtree.nim]: a300d24da3482bc1a548467654b6bc1b98966fbc

**Comment by arnetheduck** [comment]
Good to go from my side!

**Review by potuz — DISMISSED** [review]

**Review by potuz — APPROVED** [review]

**Pull request merged**